pkg/builder/cluster/convert/convert.go

```diff
@@ -105,87 +106,169 @@ var (
 		return nil, fmt.Errorf("Unexpectedly few arguments to git source container: %v", git.Args)
 	}
+
+	name := strings.TrimPrefix(git.Name, gitSource+"-")
+
+	// undo the suffix of index if name is empty
+	if _, err := strconv.Atoi(name); err == nil {
+		name = ""
+	}
+
+	var path string
+	// if `--path ` is set in args extract targetpath variable
+	if len(git.Args) == 6 {
+		path = git.Args[5]
+	}
+
 	// Now undo what we did above
 	return &v1alpha1.SourceSpec{
+		Name: name,
 		Git: &v1alpha1.GitSourceSpec{
 			Url:      git.Args[1],
 			Revision: git.Args[3],
 		},
+		TargetPath: path,
 	}, nil
 }
 
-func gcsToContainer(gcs *v1alpha1.GCSSourceSpec) (*corev1.Container, error) {
+func gcsToContainer(source v1alpha1.SourceSpec, index int) (*corev1.Container, error) {
+	gcs := source.GCS
 	if gcs.Location == "" {
 		return nil, validation.NewError("MissingLocation", "gcs sources are expected to specify a Location, got: %v", gcs)
 	}
+	args := []string{"--type", string(gcs.Type), "--location", gcs.Location}
+	// dest_dir is the destination directory for GCS files to be copies"
+	if source.TargetPath != "" {
+		args = append(args, "--dest_dir", filepath.Join(workspaceDir, source.TargetPath))
+	}
+
+	// source name is empty then use `build-step-gcs-source` name
+	containerName := initContainerPrefix + gcsSource + "-"
+
+	// update container name to include `name` as suffix
+	if source.Name != "" {
+		containerName = containerName + source.Name
+	} else {
+		containerName = containerName + strconv.Itoa(index)
+	}
+
 	return &corev1.Container{
-		Name:         initContainerPrefix + gcsSource,
+		Name:         containerName,
 		Image:        *gcsFetcherImage,
-		Args:         []string{"--type", string(gcs.Type), "--location", gcs.Location},
+		Args:         args,
 		VolumeMounts: implicitVolumeMounts,
 		WorkingDir:   workspaceDir,
 		Env:          implicitEnvVars,
 	}, nil
 }
 
 func containerToGCS(source corev1.Container) (*v1alpha1.SourceSpec, error) {
-	var sourceType, location string
+	var sourceType, location, name, path string
 	for i, a := range source.Args {
 		if a == "--type" && i < len(source.Args) {
 			sourceType = source.Args[i+1]
 		}
 		if a == "--location" && i < len(source.Args) {
 			location = source.Args[i+1]
 		}
+		if a == "--dest_dir" && i < len(source.Args) {
+			path = strings.TrimPrefix(source.Args[i+1], workspaceDir+"/")
+		}
+	}
+	// Undo the suffixing to extract source name
+	name = strings.TrimPrefix(source.Name, gcsSource+"-")
+
+	// undo the suffix of index if name is integer
+	if _, err := strconv.Atoi(name); err == nil {
+		name = ""
 	}
+
 	return &v1alpha1.SourceSpec{
+		Name: name,
 		GCS: &v1alpha1.GCSSourceSpec{
 			Type:     v1alpha1.GCSSourceType(sourceType),
 			Location: location,
 		},
+		TargetPath: path,
 	}, nil
 }
 
-func customToContainer(source *corev1.Container) (*corev1.Container, error) {
+func customToContainer(source *corev1.Container, name string) (*corev1.Container, error) {
```

Comment:
How are custom containers supposed to consume the value of `targetPath`? 

It seems like we should either inject the value somehow (into an env?) or else reject targetPath != ""

Reply:
Thats a valid point. Currently targetPath is not used in custom sources so we should raise and document it in targetPath.

Reply:
Fixed it and added test case for this.